### PR TITLE
Molpro

### PR DESCRIPTION
**Pull request by valtandor**
Changes to Molpro easyblock to accommodate precompiled binaries

### CODE REVIEW

**Comment by valtandor** [comment]
Yes

File: easybuild/easyblocks/m/molpro.py [hunk 1]
```diff
@@ -67,85 +88,95 @@ def configure_step(self):
...
+                sys.stdout.write(line)
 
-        # reread CONFIG and log contents
-        cfgtxt = read_file(cfgfile)
-        self.log.info("Contents of CONFIG file:\n%s", cfgtxt)
+            # reread CONFIG and log contents
+            cfgtxt = read_file(cfgfile)
+            self.log.info("Contents of CONFIG file:\n%s", cfgtxt)
+
+    def build_step(self):
```
Comment (valtandor) [inline easybuild/easyblocks/m/molpro.py]: @boegel: You suggested build_step probably isn't necessary on account of the superclass method being chosen by default. How does that work given that we only want to run build_step if a compilation needs to be done?

Comment (boegel) [inline easybuild/easyblocks/m/molpro.py]: I was wrong, you need this check there

but, it should be this:

```python
super(EB_Molpro, self).build_step()
```

so, please retest your easyblock after making changes?

File: easybuild/easyblocks/m/molpro.py [hunk 2]
```diff
@@ -51,6 +66,12 @@ def __init__(self, *args, **kwargs):
         self.cleanup_token_symlink = False
         self.license_token = os.path.join(os.path.expanduser('~'), '.molpro', 'token')
 
+    def extract_step(self):
+        if self.cfg['precompiled_binaries']:
+            Binary.extract_step(self)
+        else:
+            super.extract_step(self)
```
Comment (boegel) [inline easybuild/easyblocks/m/molpro.py]: fix the `super` call?

File: easybuild/easyblocks/m/molpro.py [hunk 3]
```diff
@@ -32,15 +32,30 @@
...
+class EB_Molpro(ConfigureMake, Binary):
     """Support for building and installing Molpro."""
 
+    @staticmethod
+    def extra_options():
+        """Define custom easyconfig parameters for Molpro."""
+        # Combine extra variables from Binary and ConfigureMake easyblocks as
+        # well as those needed for Molpro specifically
+        extra_vars = Binary.extra_options()
+        extra_vars = super.extra_options(extra_vars)
```
Comment (boegel) [inline easybuild/easyblocks/m/molpro.py]: `super` won't work here because this is a `staticmethod`

File: easybuild/easyblocks/m/molpro.py [hunk 4]
```diff
@@ -51,6 +66,12 @@ def __init__(self, *args, **kwargs):
         self.cleanup_token_symlink = False
         self.license_token = os.path.join(os.path.expanduser('~'), '.molpro', 'token')
 
+    def extract_step(self):
+        if self.cfg['precompiled_binaries']:
+            Binary.extract_step(self)
+        else:
+            ConfigureMake.extract_step(self)
```
Comment (valtandor) [inline easybuild/easyblocks/m/molpro.py]: The superclass doesn't work for this - it has to be ConfigureMake explicitly.

**Comment by valtandor** [comment]
I think this should be good now.

**Comment by boegel** [comment]
@wpoely86: can you give this another look?

File: easybuild/easyblocks/m/molpro.py [hunk 5]
```diff
@@ -67,85 +88,95 @@ def configure_step(self):
                     raise EasyBuildError("Failed to create symlink for license token at %s", self.license_token)
 
             else:
-                # no license token available, so fail early
-                raise EasyBuildError("No license token found at either %s or via 'license_file'", self.license_token)
+                self.log.warning("No licence token found at either {0} or via 'license_file'".format(self.license_token))
```
Comment (wpoely86) [inline easybuild/easyblocks/m/molpro.py]: why not fail anymore?

File: easybuild/easyblocks/m/molpro.py [hunk 6]
```diff
@@ -67,85 +88,95 @@ def configure_step(self):
...
+        
+        # Only do the rest of the configuration if we're building from source 
+        if not self.cfg['precompiled_binaries']:
```
Comment (wpoely86) [inline easybuild/easyblocks/m/molpro.py]: delete this empty line.

File: easybuild/easyblocks/m/molpro.py [hunk 7]
```diff
@@ -67,85 +88,95 @@ def configure_step(self):
...
-        # determine full installation prefix
-        prefix_regex = re.compile('^PREFIX=(.*)$', re.M)
-        res = prefix_regex.search(cfgtxt)
-        if res:
-            self.full_prefix = res.group(1)
-            self.log.debug("Found full installation prefix: %s", self.full_prefix)
-        else:
-            raise EasyBuildError("Failed to determine full installation prefix")
+            # determine original LAUNCHER value
+            launcher_regex = re.compile('^LAUNCHER=(.*)$', re.M)
```
Comment (wpoely86) [inline easybuild/easyblocks/m/molpro.py]: this reads all lines following `LAUNCHER=`?

Comment (wpoely86) [inline easybuild/easyblocks/m/molpro.py]: a no, why the `re.M`?

File: easybuild/easyblocks/m/molpro.py [hunk 8]
```diff
@@ -67,85 +88,95 @@ def configure_step(self):
...
+            launcher = self.toolchain.mpi_cmd_for('%x', self.cfg['parallel'])
+            launcher = launcher.replace(' %s' % self.cfg['parallel'], ' %n')
 
-        # patch CONFIG file to change LAUNCHER definition, in order to avoid having to start mpd
-        for line in fileinput.input(cfgfile, inplace=1, backup='.orig'):
-            line = re.sub(r"^(LAUNCHER\s*=\s*).*$", r"\1 %s" % launcher, line)
-            sys.stdout.write(line)
+            # patch CONFIG file to change LAUNCHER definition, in order to avoid having to start mpd
+            for line in fileinput.input(cfgfile, inplace=1, backup='.orig'):
+                line = re.sub(r"^(LAUNCHER\s*=\s*).*$", r"\1 %s" % launcher, line)
```
Comment (wpoely86) [inline easybuild/easyblocks/m/molpro.py]: why is the regex for `LAUNCHER` different then the one on L134 ?

File: easybuild/easyblocks/m/molpro.py [hunk 9]
```diff
@@ -154,15 +185,45 @@ def install_step(self):
...
+                # Questions whose text may be matched as a regular expression
+                stdqa = {
+                        r"Enter installation directory for executable files \[.*\]\n": "{0}/bin".format(self.installdir),
+                        r"Enter installation directory for library files \[.*\]\n": "{0}/lib".format(self.installdir),
+                        r"directory .* does not exist, try to create [Y]/n\n": '',
+                }
+                run_cmd_qa(cmd, qa=qa, std_qa=stdqa, log_all=True, simple=True)
+
+        else:
+
```
Comment (wpoely86) [inline easybuild/easyblocks/m/molpro.py]: cut down on the empty lines.

File: easybuild/easyblocks/m/molpro.py [hunk 10]
```diff
@@ -154,15 +185,45 @@ def install_step(self):
...
+            if os.path.isfile(self.license_token):
+                run_cmd("make tuning")
```
Comment (wpoely86) [inline easybuild/easyblocks/m/molpro.py]: what does this do? add a comment.

**Comment by wpoely86** [comment]
looks good

File: easybuild/easyblocks/m/molpro.py (hunk 5 above)
Comment (valtandor) [inline easybuild/easyblocks/m/molpro.py]: This change was made to accommodate cases in which one installation of Molpro might serve several users, each of which brings his own licence token to the table, so to speak. Requiring a token at installation time (when the HPC facility might not have its own Molpro licence but is installing on behalf of users) breaks this approach.

File: easybuild/easyblocks/m/molpro.py (hunk 7 above)
Comment (valtandor) [inline easybuild/easyblocks/m/molpro.py]: I guess whoever first wrote the molpro easyblock can comment on that.

File: easybuild/easyblocks/m/molpro.py (hunk 8 above)
Comment (valtandor) [inline easybuild/easyblocks/m/molpro.py]: I guess whoever first wrote the molpro easyblock can comment on that.

File: easybuild/easyblocks/m/molpro.py (hunk 10 above)
Comment (valtandor) [inline easybuild/easyblocks/m/molpro.py]: I guess whoever first wrote the molpro easyblock can comment on that.

**Comment by boegel** [comment]
Jenkins: ok to test

**Comment by boegel** [comment]
@valtandor: I took care of some minor style fixes in https://github.com/valtandor/easybuild-easyblocks/pull/2

**Comment by boegel** [comment]
continued in #807, where I also added support for Molpro 2015.x and use `apply_regex_substitutions`

Thanks @valtandor!

**Comment by boegel** [comment]
merged via #807 

The last "sync with develop" commit is confusing GitHub a bit, but everything is there, so closing this.

Thanks again @valtandor!

**Pull request closed without merging**